---
title: "Lake Erie HABs 16 data"
output:
  html_document:
    toc: true
    theme: united
---


# Libraries
```{r load libraries,warning=FALSE,message=FALSE}
#Load libraries
library(phyloseq)
packageVersion("phyloseq")
library(Biostrings)
library(ggplot2)
library(vegan)
library(plyr)
library(scales)
library(grid)
library(reshape2)
library(gridExtra)
library(RColorBrewer)


# Set working directory
setwd("~/habs/genomics/data/")

```

# Data import
```{r mothur import}

# Import mothur files and sample metadata
sharedfile = "mothur/allhabs.shared"
taxfile = "mothur/allhabs.taxonomy"
mapfile = "best_metadata.csv"
 
mothurdata = import_mothur(mothur_shared_file = sharedfile,
                             mothur_constaxonomy_file = taxfile)

# Add the OTU number as a column in the taxonomy file
tax_table(mothurdata) <- cbind(tax_table(mothurdata),
                               row.names(tax_table(mothurdata)))

# Rename the taxonomy columns
colnames(tax_table(mothurdata)) <- c("Kingdom","Phylum","Class",
                                       "Order","Family","Genus","Species")
  
# Import sample metadata
map <- read.csv(mapfile)
map <- sample_data(map)
rownames(map) <- map$SampleID
  
# Merge mothurdata object with sample metadata
moth_merge = merge_phyloseq(mothurdata,map)

# Filter out non-samples (i.e. water,mock) and samples from intensive cruises.
# Also prune out taxa which were only present in removed samples
moth_good <- subset_samples(moth_merge,Type=="sample")
moth_good <- prune_taxa(taxa_sums(moth_good)>0,moth_good)

# Filter out non-bacteria, chloroplasts and mitochondria
moth_good <- subset_taxa(moth_good, Kingdom == "Bacteria")
moth_good <- subset_taxa(moth_good, Class != "Chloroplast")
moth_good <- subset_taxa(moth_good, Family!="mitochondria")



```

```{r representative sequences,echo=FALSE,include=FALSE}

# Import representative sequences from mothur
ref.seq <- readDNAStringSet("mothur/mothur_repseq.fasta")

# Split on tab to seperate sequence header and OTU number
ref1 <- data.frame(do.call('rbind',
                           strsplit(as.character(names(ref.seq)),'\t',
                                    fixed=TRUE)))
# Split on | to seperate OTU number
ref2 <- data.frame(do.call('rbind',
                           strsplit(as.character(ref1$X2),"|",
                                    fixed=TRUE)))

# Make the OTU names the rownames for the reference sequences
names(ref.seq)<-ref2$X1

# Turn into refseq object
ref.seq <- refseq(ref.seq)


```

```{r functions,echo=FALSE}
### all the functions####

##### FUNCTIONS ########

# Scale reads using mcmurdie holmes method to a given library size of n
scale_reads <- function(physeq,n){
  physeq.scale <- transform_sample_counts(physeq, function(x) {(n*x/sum(x))})
  otu_table(physeq.scale) = floor(otu_table(physeq.scale))
  physeq.scale = prune_taxa(taxa_sums(physeq.scale)>0,physeq.scale)
  return(physeq.scale)
}

# Order dates correctly in a data frame
order_dates <- function(df){
  df$Date <- factor(df$Date, levels = c("5/27","6/10","6/16","6/30","7/8","7/14","7/21",
                                        "7/29","8/4","8/11","8/18","8/25","9/2","9/8","9/15",
                                        "9/23","9/29","10/6","10/15","10/20","10/27","11/3"))
  return(df)
}

# Function to run adonis test on a distance object and a categorical variable from a metadata data frame
doadonis <- function(physeq, category) {
  physeq.scale <- scale_reads(physeq,min(sample_sums(physeq)))
  bdist<-phyloseq::distance(physeq.scale,"bray")
  col = as(sample_data(physeq),"data.frame")[,category]
  adonis.bdist=adonis(bdist~col)
  print(adonis.bdist) 
  paste("The Homogeneity of dispersion \n")
 

 # Homogeneity of dispersion test
  betatax=betadisper(bdist,col)
  p=permutest(betatax)
  print("The homogeneity of dispersion:") 
  print(p$tab)
}  


# Function to make a pcoa plot with bray-cutis distance where color variable is discrete
make_pcoa <- function(physeq){
  physeq.scale <- scale_reads(physeq,min(sample_sums(physeq)))
  sample_data(physeq.scale) <- order_dates(sample_data(physeq.scale))
  set.seed(1)
  physeq.bray.nmds = ordinate(physeq.scale,method="PCoA","bray")
}  
  

# This function takes a phyloseq object, agglomerates OTU's to the desired taxonomic rank, prunes out OTUs below a certain relative proportion in a sample (ie 1% ) and melts the phyloseq object into long format.
transform_and_melt <- function(physeq, taxrank, prune){
  
  # Agglomerate all otu's by phylum level
  physeq_taxrank <- tax_glom(physeq,taxrank = taxrank)

  # Create a new phyloseq object which removes taxa from each sample below the prune parameter 
  physeq_taxrank.prune <- transform_sample_counts(physeq_taxrank, 
                                              function(x){x/sum(x)})
  otu_table(physeq_taxrank.prune)[otu_table(physeq_taxrank.prune) < prune] <- 0 
  physeq_taxrank.prune <- prune_taxa(taxa_sums(physeq_taxrank.prune) > 0,
                                 physeq_taxrank.prune)
  
  # Set all samples from 8/11 to 0 because of bad sampling on that date
  w<-which(sample_data(physeq_taxrank.prune)$Date == "8/11")
  otu_table(physeq_taxrank.prune)[,w] <- 0
  
  # Melt into long format and sort by samplenum and phylum
  physeq_taxrank.long <- psmelt(physeq_taxrank.prune)
  physeq_taxrank.long <- arrange(physeq_taxrank.long,Samplenum,Phylum)


  # Reorder factor levels for dates 
  physeq_taxrank.long <- order_dates(physeq_taxrank.long)
  
  # Reorder factor levers for stations
  physeq_taxrank.long$Station <- factor(physeq_taxrank.long$Station,
                                        levels=c("WE2","WE12","WE4"))
  
  # Reorder factor levels for Fraction. Change Fraction factor names
  physeq_taxrank.long$Fraction <- factor(physeq_taxrank.long$Fraction,
                                        levels=c("CNA","100LTR","53LTR","3NA","22NA"))
  
  levels(physeq_taxrank.long$Fraction) <- c("Full", "100um", "53um","3um","0.22um")
  
  # Return long data frame
  return(physeq_taxrank.long)
}

# This function takes  a data frame in long format (such as the output from transform_and_melt) and produces a stacked barplot of community composition.   
make_tax_barplot<-function(df,x,y,tax,facet,title,colors,xlab,ylab,relative,outline,guide){
  ggplot(df,aes_string(x=x,y=y,fill=tax)) +
        facet_grid(reformulate(facet))+
        geom_bar(stat="identity",show_guide=guide)+
    scale_fill_manual(values = colors) + 
    scale_x_discrete(breaks=c("6/10","7/8",
                              "8/4","9/2","10/6","11/3"),
                      labels=c("Jun", "Jul",
                                "Aug", "Sep",  "Oct","Nov"),
                     drop=FALSE
                     )+
    theme(axis.title.x = element_text(size=16,face="bold"),
          axis.text.x = element_text(angle=50, colour = "black", vjust=1, hjust = 1, size=13),
          axis.text.y = element_text(colour = "black", size=10),
          axis.title.y = element_text(face="bold", size=16),
          plot.title = element_text(size = 18),
          legend.title = element_text(size=14),
          legend.text = element_text(size = 13),
          legend.position="right",
          strip.text.x = element_text(size=16, face="bold"),
          strip.text.y = element_text(size=16, face="bold"),
          strip.background = element_rect(color="white",size=2,fill=NA),
          panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.border = element_rect(colour = "black",fill=NA,size=1.5),
          panel.margin = unit(1, "lines"),
          panel.background = element_rect(fill="#a8a8a8")
      ) +
    guides(fill = guide_legend(reverse= TRUE,keywidth=1,keyheight=1))+
    xlab(xlab)+
    ylab(ylab)+
    ggtitle(title)+
    if (relative){
          geom_bar(stat="identity",position="fill", colour=outline,show_guide=FALSE) 
        } else {
          geom_bar(stat="identity",colour=outline,show_guide=FALSE) 
    }  
        
}

# Credit to Hadley Wickham for this function which multiplots with one legend
grid_arrange_shared_legend <- function(...) {
    plots <- list(...)
    g <- ggplotGrob(plots[[1]] + theme(legend.position="bottom"))$grobs
    legend <- g[[which(sapply(g, function(x) x$name) == "guide-box")]]
    lheight <- sum(legend$height)
    grid.arrange(
        do.call(arrangeGrob, lapply(plots, function(x)
            x + theme(legend.position="none"))),
        legend,
        ncol = 1,
        heights = unit.c(unit(1, "npc") - lheight, lheight))
}

# Plot multiple ggplot outputs
multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {

  # Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)

  numPlots = length(plots)

  # If layout is NULL, then use 'cols' to determine layout
  if (is.null(layout)) {
    # Make the panel
    # ncol: Number of columns of plots
    # nrow: Number of rows needed, calculated from # of cols
    layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                    ncol = cols, nrow = ceiling(numPlots/cols))
  }

 if (numPlots==1) {
    print(plots[[1]])

  } else {
    # Set up the page
    grid.newpage()
    pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))

    # Make each plot, in the correct location
    for (i in 1:numPlots) {
      # Get the i,j matrix positions of the regions that contain this subplot
      matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))

      print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                      layout.pos.col = matchidx$col))
    }
  }
}



```

# Sample summary
A lot of the 100um samples have very low read counts because a majority of the reads were chloroplasts. Before removing non-bacteria, all samples had at least 10,000 reads, afterwards many of the 100L samples had as few as 500 reads. This is important to keep in mind for beta diversity analyses, because to include all the samples we need to really throw out a lot of data to normalize the sequencing depth.

```{r,echo=FALSE}
# Histogram of sample read counts
ggplot(data.frame(sum=sample_sums(moth_good)), aes(sum)) + 
  geom_histogram(colour="white", fill="indianred") +
  ggtitle("Sample read counts") + 
  xlab("total sequences")

# mean, max and min of sample read counts
mins<-min(sample_sums(moth_good))
means<-mean(sample_sums(moth_good))
maxs<-max(sample_sums(moth_good))

```
The minimum sample read count is $mins
The mean sample read count is $means
The max sample read count is $maxs


# Pigment/toxin/nutrient data

## Data import

```{r}
# Import metadata file with nutrients, pigments and toxin
nutrient <- read.csv("other/habs_nutrients.csv")

# Format nutrient data frame
nutrient <- order_dates(nutrient)
nutrient$Station <- factor(nutrient$Station, levels=c("WE2","WE4","WE12"))
```

## Nutrient time-series
```{r}

# Nitrate vs ParMC
plot(nutrient$Nitrate, nutrient$ParMC, xlab="Nitrate", ylab="ParMC")

# Nitrate vs ParMC until 8/11
Nitrate_early <- subset(nutrient, Date %in% c("5/27","6/10","6/16","6/30","7/8",
                                              "7/14","7/21","7/29","8/4","8/11"))

ggplot(Nitrate_early,aes(x=Nitrate,y=ParMC, group=Station)) +
  geom_point() +
  scale_y_log10() + 
  
# Nitrate vs relative abundance of oligotype TG in 100um fraction
                   
```


## Figure 1 
```{r}

# Melt nutrient data frame on variables phycocyanin and parmc
nutrient.melt <- melt(nutrient,measure.vars = c("Phycocyanin","ParMC"))

# Plot facets of stations by phyco/parmc
theme_set(theme_classic())
ggplot(nutrient.melt,aes(x=Date,y=value,group=Station,color=Station)) +
      facet_wrap(variable~Station,scales="free_y") +
      scale_x_discrete(breaks=c("6/10","7/8","8/4","9/2","10/6","11/3"),
                      labels=c("Jun", "Jul","Aug", "Sep", "Oct", "Nov"),
                      drop=FALSE
                      ) +
      geom_line(size=.8,show_guide=FALSE) +
      geom_point(size=1.8,show_guide=FALSE) +
      ggtitle("") +
      xlab("") + 
      ylab("") +       
      scale_color_manual(values=c("#E96446","#52361b","#87CEFA")) +
      theme(axis.text.x = element_text(angle=45, vjust=1, hjust = 1, size=12, face="bold"),  
            strip.text.x = element_blank(),
            strip.background = element_blank(),
            panel.border = element_rect(colour = "black",fill=NA,size=1.5)
        )

```

# Figure 2: Nutrients
```{r plot unpooled, fig.height=10,fig.width=10, warning=FALSE}

nutplot_unpooled <- function(df,nutrient,title){
        ggplot(df,
                aes_string(x="Date",y=nutrient,group="Station",shape="Station",color="Station"))+
         geom_line(size=.8)+
         geom_point(size=1.8)+
         ggtitle(title)+ 
         xlab("") + ylab("ug/L")+       
         scale_color_manual(values=c("#E96446","#52361b","#87CEFA"))+
         theme(axis.text.x = element_text(angle=45,vjust=1, hjust = 1, size=9, face="bold"),     
               axis.title.y = element_text(size=14),
               plot.title = element_text(size=16,face="bold"))
}

# Plot phosphorus
p_SRP <- nutplot_unpooled(nutrient,"SRP","SRP")
                                  
# Plot nitrate
p_NO3 <- nutplot_unpooled(nutrient,"Nitrate","Nitrate")

# Plot Ammonia
p_NH4 <- nutplot_unpooled(nutrient,"Ammonia","Ammonia")

# Plot H202 
p_h202 <- nutplot_unpooled(nutrient,"H2O2","H2O2")

# Plot Chla
p_chla <-nutplot_unpooled(nutrient,"Chla","Chlorophyll A")

# Plot temp
p_temp <- nutplot_unpooled(nutrient,"Temp","Temperature")


# multiplot
theme_set(theme_classic())
grid_arrange_shared_legend(p_NO3,p_chla,p_NH4,p_temp,p_SRP,p_h202)

```

# Figure S4
```{r}

# Plot secchi
p_secchi <-  nutplot_unpooled(othernuts,"Secchi","Secchi")

# Plot turbidity
p_turb <-  nutplot_unpooled(othernuts,"Turbidity","Turbidity")

# Plot DOC
p_doc <- nutplot_unpooled(nutrient,"DOC","DOC") 

# Plot PP
p_pp <- nutplot_unpooled(nutrient,"PP","Particulate P") 

# Plot PON
p_pon<- nutplot_unpooled(nutrient,"PON","Particulate Org. N") 

theme_set(theme_classic())
grid_arrange_shared_legend(p_secchi,p_turb,p_pp,p_pon)

```

# Figure 4: Microcystis Oligotypes

```{r,warning=FALSE,fig.height=6,fig.width=14}

# Read in oligotype count data
o <- read.csv("oligotype/oligo_formatted_all3.csv", sep= ",")

# Remove unnecessary dates and fractions
o.sub <- subset(o, !(Date %in% c("5/27", "6/10", "6/16", "6/30")) & Fraction != "22NA")
o.sub$Date <- factor(o.sub$Date, levels=c("7/8","7/14","7/21",
                                        "7/29","8/4","8/11","8/18","8/25","9/2","9/8","9/15",
                                        "9/23","9/29","10/6","10/15","10/20","10/27","11/3"))

# Melt the oligotypes 
oligo.melt<-melt(value.name = "count",
                 variable.name = "Oligotype",
                 id.vars=c("SampleID", "Station", "Date", "Fraction"),
                 o.sub)

# Reorder factor levers for stations
oligo.melt$Station <- factor(oligo.melt$Station,
                                      levels=c("WE2", "WE12", "WE4"))

# Reorder factor levels for Fraction. Change Fraction factor names
oligo.melt$Fraction <- factor(oligo.melt$Fraction,
                                       levels=c("CNA", "100LTR", "53LTR", "3NA"))
  
levels(oligo.melt$Fraction) <- c("Full", "100um", "53um", "3um")

# Plot
theme_set(theme_classic())

ggplot(oligo.melt, 
        aes_string(x="Date", y="count", fill="Oligotype")) +
        facet_wrap(Fraction~Station,ncol=3) +
        scale_x_discrete(breaks = c("7/8", "8/4", "9/2", "10/6", "11/3"),
                         labels = c("Jul", "Aug", "Sep",  "Oct", "Nov"),
                         drop = FALSE
                         ) +
        geom_bar(stat="identity") +
        scale_fill_manual(breaks = c("TG","CG","TT"),
                          values = c("#CA3543","#1F9EAE","#694B3D","#e9e9e9")) + 
        theme(axis.title.x = element_text(size=16,face="bold"),
              axis.text.x = element_text(angle=50, colour = "black", 
                                         vjust=1, hjust = 1, size=12,face="bold"),
              axis.text.y = element_text(colour = "black", size=12),
              axis.title.y = element_text(face="bold", size=16),
              legend.title = element_text(size=18),
              legend.text = element_text(size = 16),
              strip.text.x = element_text(size=18, face="bold"),
              strip.text.y = element_text(size=16, face="bold"),
              strip.background = element_blank(),
              panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              panel.background = element_rect(fill=NA, color="black"),
              axis.line = element_line(colour = "black")
          ) +
       guides(fill = guide_legend(reverse = TRUE, keywidth=1, keyheight=1)) +   
       xlab("") +
       ylab("Relative Abundance \n") +
       ggtitle("")

```

## Anabaena
```{r}

# Read in oligotype count data
ana <- read.csv("oligotype/anabaena_counts.txt",sep="\t")

oligodat<-data.frame(sample_data(moth_good)[,c(1,3,4,8)],Sums=sample_sums(moth_good))
ana.merge<-merge(ana,oligodat,by.x="samples",by.y="SampleID")

ana.scaled<-ana.merge[,2:6]/ana.merge[,10]

ana.format<-data.frame(ana.scaled,ana.merge[,c(1,7:9)])
ana.format<-subset(ana.format, !(Date %in% c("11/3")) &Fraction != "22NA")


# Melt the oligotypes 
ana.melt<-melt(value.name = "count",
                 variable.name = "Oligotype",
                 id.vars=c("samples","Station","Date","Fraction"),
                 ana.format)

# Order dates
ana.melt<-order_dates(ana.melt)

# Reorder factor levers for stations
ana.melt$Station <-factor(ana.melt$Station,
                                      levels=c("WE2","WE12","WE4"))

# Reorder factor levels for Fraction. Change Fraction factor names
ana.melt$Fraction <- factor(ana.melt$Fraction,
                                       levels=c("CNA","100LTR","53LTR","3NA","22NA"))
  
levels(ana.melt$Fraction) <- c("Full", "100um","53um","3um",".22um")

# Plot
theme_set(theme_classic())

ggplot(ana.melt,aes_string(x="Date",y="count",fill="Oligotype")) +
        facet_grid(Fraction~Station,scales = "free_y")+
        geom_bar(stat="identity")+
    scale_fill_manual(values = c("grey26","chartreuse3", "darkorange","royalblue", "red", 
                               "cyan2", "darkgreen")) + 
    scale_x_discrete(breaks=c("6/10","7/8",
                              "8/4","9/2","10/6","11/3"),
                      labels=c("Jun", "Jul",
                                "Aug", "Sep",  "Oct","Nov"),
                     drop=FALSE
                     )+
    theme(axis.title.x = element_text(size=16,face="bold"),
          axis.text.x = element_text(angle=50, colour = "black", 
                                     vjust=1, hjust = 1, size=12,face="bold"),
          axis.text.y = element_text(colour = "black", size=12),
          axis.title.y = element_text(face="bold", size=16),
          plot.title = element_text(size = 18),
          legend.title = element_text(size=18),
          legend.text = element_text(size = 16),
          legend.position="right",
          strip.text.x = element_text(size=18, face="bold"),
          strip.text.y = element_text(size=16, face="bold"),
          strip.background = element_blank(),
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          panel.background = element_rect(fill=NA, color="black"),
          axis.line = element_line(colour = "black")
        )+
    guides(fill = guide_legend(reverse= TRUE,keywidth=1,keyheight=1))+   
    xlab("")+
    ylab("Relative Abundance \n")+
    ggtitle("")

```

## Pseudanabaena

```{r}

# Read in oligotype count data
pseu<-read.csv("oligotype/pseudanabaena_counts",sep="\t")

oligodat<-data.frame(sample_data(moth_good)[,c("Date","Fraction","Station","SampleID")],Sums=sample_sums(moth_good))
pseu.merge<-merge(pseu,oligodat,by.x="samples",by.y="SampleID")

pseu.scaled<-pseu.merge[,2:5]/pseu.merge[,9]

pseu.format<-data.frame(pseu.scaled,pseu.merge[,c(1,6:8)])
pseu.format<-subset(pseu.format, Fraction != "22NA")


# Melt the oligotypes 
pseu.melt<-melt(value.name = "count",
                 variable.name = "Oligotype",
                 id.vars=c("samples","Station","Date","Fraction"),
                 pseu.format)

# Order dates
pseu.melt<-order_dates(pseu.melt)

# Reorder factor levers for stations
pseu.melt$Station <-factor(pseu.melt$Station,
                                      levels=c("WE2","WE12","WE4"))

# Reorder factor levels for Fraction. Change Fraction factor names
pseu.melt$Fraction <- factor(pseu.melt$Fraction,
                                       levels=c("CNA","100LTR","53LTR","3NA","22NA"))
  
levels(pseu.melt$Fraction) <- c("Full", "100um","53um","3um",".22um")

# Plot
theme_set(theme_bw())
scale_fill_manual(values=c( "violetred", 
                               "deepskyblue","blue4", "chartreuse3"))

ggplot(pseu.melt,aes_string(x="Date",y="count",fill="Oligotype")) +
        facet_grid(Fraction~Station,scales = "free_y")+
        geom_bar(stat="identity")+
    scale_fill_manual(values = c("violetred", 
                               "deepskyblue","blue4", "chartreuse3")
                      ) + 
    scale_x_discrete(breaks=c("6/10","7/8",
                              "8/4","9/2","10/6","11/3"),
                      labels=c("Jun", "Jul",
                                "Aug", "Sep",  "Oct","Nov"),
                     drop=FALSE
                     ) +
    theme(axis.title.x = element_text(size=16,face="bold"),
          axis.text.x = element_text(angle=50, colour = "black", 
                                     vjust=1, hjust = 1, size=12,face="bold"),
          axis.text.y = element_text(colour = "black", size=12),
          axis.title.y = element_text(face="bold", size=16),
          plot.title = element_text(size = 18),
          legend.title = element_text(size=18),
          legend.text = element_text(size = 16),
          legend.position="right",
          strip.text.x = element_text(size=18, face="bold"),
          strip.text.y = element_text(size=16, face="bold"),
          strip.background = element_blank(),
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          panel.background = element_rect(fill=NA, color="black"),
          axis.line = element_line(colour = "black")
        )+
    guides(fill = guide_legend(reverse= TRUE,keywidth=1,keyheight=1))+   
    xlab("")+
    ylab("Relative Abundance \n")+
    ggtitle("")

```

# Full Community Analysis

## Phylum composition across stations and filters
```{r composition barplot, fig.height=10,fig.width=14}


# Colors from ColorBrewer Paired, 12
phylum.colors <- c(Acidobacteria="#4D4D4D",Actinobacteria="#ff7f00",
                   Armatimonadetes="#ffff99", Bacteroidetes="#6a3d9a"
                   ,Candidate_division_SR1 ="#b15928" ,Chlorobi="#cab2d6", 
                   Chloroflexi="#B2DF8A",Chloroplast="darkgreen",
                   Cyanobacteria="#33A02C",Gemmatimonadetes="#fdbf6f",
                   Planctomycetes="#A6CEE3",Proteobacteria="#1F78B4", 
                   unclassified="#fb9a99", Verrucomicrobia ="#e31a1c",
                   Nitrospirae="yellow","BD1-5"="white",Firmicutes="blue") 


# Website colors
phylum.colors<- c("#CBD588", "#5F7FC7", "orange","#DA5724",
                  "#89C5DA","#508578" , "#CD9BCD",
                  "#74D944", "#D7C1B1", "#AD6F3B", "#673770","#D14285","#689030",
                  "#6DDE88", "#652926", "#7FDCC0", "#C84248", "#8569D5", "#5E738F",
                  "#D1A33D", "#8A7C64", "#599861")


# Transform to long format and prune out phyla below 2% in each sample
moth.long<-transform_and_melt(physeq = moth_good,taxrank = "Phylum",prune = .05)

# Plot 
make_tax_barplot(df = moth.long,x = "Date", y="Abundance",
                 tax = "Phylum",
                 facet = "Fraction~Station",
                 title = "",
                 colors=phylum.colors,
                 xlab="",
                 ylab="Relative Abundance\n",
                 outline="black",
                 relative=TRUE,
                 guide=TRUE)


```

## Cyanobacteria composition
```{r cyanobacteria,echo=FALSE,include=FALSE}


# Create a new phyloseq object called cyano where the cyano b
cyano<-moth_good
t<-data.frame(tax_table(cyano))
cyanotax <- paste(t$Genus, t$Species)

cy_taxonomy<-cbind(tax_table(cyano),cyanotax)

# For each row in the taxonomy table
for  (i in 1:nrow(cy_taxonomy)){ 
   if (cy_taxonomy[i,2] != "Cyanobacteria") {
      cy_taxonomy[i,2]<-"Z"
  } else {
    cy_taxonomy[i,2]<-cy_taxonomy[i,6]
  }
}


cy_taxonomy_phyloseq <-tax_table(as.matrix(cy_taxonomy))

tax_table(cyano)<-cy_taxonomy_phyloseq

cyano<-subset_samples(cyano,Fraction!="22NA")


cyano.long <- transform_and_melt(physeq = cyano,taxrank = "Phylum",prune = .02)

cyano.long<-arrange(cyano.long,Phylum)

phylum.colors<- c("mediumorchid3","#CBD588","#DA5724",
                  "#89C5DA","orange","white","#508578" , "#CD9BCD", "#D3D93E", "#C84248",
                  "#74D944", "#D7C1B1", "#AD6F3B", "#673770","#D14285","#689030",
                  "#6DDE88", "#652926", "#7FDCC0", "#C84248", "#8569D5", "#5E738F",
                  "#D1A33D", "#8A7C64", "#599861")

theme_set(theme_grey())

make_tax_barplot(df = cyano.long,x = "Date",y="Abundance",tax = "Phylum",
                 facet = "Fraction~Station",
                 title = "",
                 colors= c("#0000e5","#009900","#8C001A","orange","mediumorchid3","white"),
                 xlab="",
                 ylab="",
                 outline="black",
                 relative=TRUE,
                 guide=FALSE)


ggplot(cyano.long,aes_string(x="Date",y="Abundance",fill="Phylum")) +
        facet_grid(Fraction~Station)+
        geom_bar(stat="identity",show_guide=FALSE)+
        geom_bar(stat="identity",position="fill",show_guide=FALSE)+
    scale_fill_manual(values = c("#0000e5","#009900",
                                 "#8C001A","orange",
                                 "mediumorchid3","white")) + 
    scale_x_discrete(breaks=c("6/10","7/8",
                              "8/4","9/2","10/6","11/3"),
                      labels=c("Jun", "Jul",
                                "Aug", "Sep",  "Oct","Nov"),
                     drop=FALSE
                     )+
    theme(axis.title.x = element_text(size=16,face="bold"),
          axis.text.x = element_text(angle=50, colour = "black", vjust=1, hjust = 1, size=13),
          axis.text.y = element_text(colour = "black", size=10),
          axis.title.y = element_text(face="bold", size=16),
          plot.title = element_text(size = 18),
          legend.title = element_text(size=14),
          legend.text = element_text(size = 13),
          legend.position="right",
          strip.text.x = element_text(size=16, face="bold"),
          strip.text.y = element_text(size=16, face="bold"),
          strip.background = element_rect(color="white",size=2,fill=NA),
          panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
          panel.border = element_rect(colour = "black",fill=NA,size=1.5),
          panel.margin = unit(1, "lines"),
          panel.background = element_rect(fill="#a8a8a8")
      ) +
    guides(fill = guide_legend(reverse= TRUE,keywidth=1,keyheight=1))+
    xlab("")+
    ylab("Relative Abundance")

  
  
  
  
  
  

grid.arrange(arrangeGrob(p_phyco,p_parmc, ncol=1),cyanobar, ncol=1, heights=c(1,1,4))



```

```{r cyano ordinations}

# Subset to just cyanobacteria
cyanos <- subset_taxa(moth_good, Phylum == "Cyanobacteria")

# Subset to just full community samples
cyanos.cna <- subset_samples(cyanos, Fraction == "CNA")

# Scale to smallest library
cyanos.scale <- scale_reads(cyanos.cna, min(sample_sums(cyanos.cna)))

# Reorder station levels
sample_data(cyanos.scale)$Station <- factor(sample_data(cyanos.scale)$Station,
                                            levels=c("WE2","WE4","WE12"))

# Reorder month levels
sample_data(cyanos.scale)$Month <- factor(sample_data(cyanos.scale)$Month,
                                          levels=c("May","June","July","August",
                                                   "September","October","November"))

# Ordinate

cyanos.pcoa <- ordinate(cyanos.scale, method="PCoA", distance = "bray")


theme_set(theme_classic())

cyanos.pcoa.plot <- plot_ordination(physeq = cyanos.scale,
                                   ordination = cyanos.pcoa,
                                   color = "Month",
                                   shape = "Station",
                                   axes = c(1,2)) + 
                  geom_point(aes(colour = Month),alpha=.7, size = 4) +
                  geom_point(colour="grey90", size = 1.5) + 
                  scale_color_manual(values=c("#a65628","red","#ffae19",
                              "#4daf4a","#1919ff","darkorchid3","magenta")) +
                  theme(plot.margin = unit(c(2,2,2,2), "lines"))





## Cyano CAP

cyanos.scale.sub<-subset_samples(cyanos.scale,!is.na(LogPhyco) & !is.na(SRP) &
                                !is.na(DOC) & !is.na(pH) & !is.na(ParMC) & !is.na(H2O2))


cap.ord <- ordinate(physeq = cyanos.scale.sub, method = "CAP",
                  distance = "bray",
                  formula = ~ ParMC + Nitrate + SRP + LogPhyco + Ammonia + DOC + pH + H2O2)

sample_data(cyanos.scale.sub)$Station <- factor(sample_data(cyanos.scale.sub)$Station,
                                            levels=c("WE2","WE4","WE12"))

# Reorder month levels
sample_data(cyanos.scale.sub)$Month <- factor(sample_data(cyanos.scale.sub)$Month,
                                              levels=c("May","June","July","August",
                                                       "September","October","November"))



cap.plot <- plot_ordination(cyanos.scale.sub, cap.ord, color="Month",axes = c(1,2)) + 
            aes(shape = Station) +
            geom_point(aes(colour = Month),alpha=.5, size = 4) +
            geom_point(colour="grey90", size = 1.5) + 
            scale_color_manual(values=c("#a65628","red","#ffae19",
                                        "#4daf4a","#1919ff",
                                        "darkorchid3","magenta"))


# Now add the environmental variables as arrows
arrowmat = vegan::scores(cap.ord, display = "bp")

# Add labels, make a data.frame
arrowdf <- data.frame(labels = rownames(arrowmat), arrowmat)

# Define the arrow aesthetic mapping
arrow_map = aes(xend = CAP1, yend = CAP2, x = 0, y = 0, 
                shape = NULL, color = NULL, 
                label = labels)

label_map = aes(x = 1.3 * CAP1, y = 1.3 * CAP2, shape = NULL, color = NULL, 
    label = labels)

arrowhead = arrow(length = unit(0.02, "npc"))

# Make a new graphic
p1 = cap.plot + 
     geom_segment(arrow_map, size = .5, 
                             data = arrowdf, color = "gray", 
                             arrow = arrowhead) + 
     geom_text(label_map, size = 4,face="bold", 
                data = arrowdf,show_guide=FALSE)


anova(cap.ord)


```

# Representative sequences
```{r representative sequences, include=FALSE}
otu.99 = transform_sample_counts(moth_good,function(x){x/sum(x)})
otu_table(otu.99)[otu_table(otu.99)<.01] <- 0 
otu.99 = prune_taxa(taxa_sums(otu.99)>0, otu.99)

writeXStringSet(refseq(otu.99),"otu.99.representatives")
                           
```

# Microcystis Associated OTUs
```{r}
library(psych)
otu<-t(otu_table(moth_good))

oligo.dat<-read.csv("oligotype/oligo_formatted.txt",sep=" ")       
oligo.dat[is.na(oligo.dat)]<-0


oligo_test <- function(df,oligotype,physeq.scale){
  oligo<-as.data.frame(df[,oligotype])
  corr<-corr.test(oligo,df[,-(1:8)],method = "pearson",adjust = "fdr")

  pvalue<-corr$p
  r <-corr$r

  sig<-which(pvalue<.05)
  otu.names<-colnames(pvalue)[sig]

  sig.tax <- data.frame(tax_table(physeq.scale))[otu.names,]
  sig.p <- pvalue[sig]
  sig.r <- r[sig]
  sig.dat<-data.frame(sig.p,sig.r,sig.tax)
  names(sig.dat)<-c("pvalue","r","Taxonomy")
  return(sig.dat)
}

heterotroph_test <- function(physeq, oligo.dat, f, prune){
  
  # Subset to specified fraction
  fraction_subset <- subset_samples(physeq,Fraction==f)
  
  # Scale to min library size and then transform to relative abundance
  fraction.scale <- scale_reads(fraction_subset,prune)
  fraction.scale <- prune_taxa(taxa_sums(fraction.scale)>0,fraction.scale)
  fraction.scale <- transform_sample_counts(physeq = fraction.scale, function(x) {(x/sum(x))})
  
  # Remove microcystis
  w<-which(rownames(otu_table(fraction.scale))=="Otu00005")
  nonmc<-data.frame(t(otu_table(fraction.scale)[-w,]))
  
  oligo.fraction <- subset(oligo.dat,Fraction=="100LTR")
  oligo.fraction$SampleID<-substr(oligo.fraction$SampleID,1,5)
  rownames(nonmc)<-substr(rownames(nonmc),1,5)
  oligo.otu.merge <- merge(oligo.fraction,nonmc,by.x="SampleID",by.y="row.names")
  
  CG.dat <- oligo_test(df = oligo.otu.merge,oligotype = "CG",physeq.scale = fraction.scale )
  CG.dat$Oligotype <- rep("CG",nrow(CG.dat))
  TG.dat <- oligo_test(df = oligo.otu.merge,oligotype = "TG",physeq.scale = fraction.scale )
  TG.dat$Oligotype <- rep("TG",nrow(TG.dat))
  hetero.dat <- rbind(CG.dat,TG.dat)
  return(hetero.dat)
  
}

f="100LTR"
hetero.100 <- heterotroph_test(physeq = moth_good,oligo.dat,f="100LTR",prune = 500)
hetero.100$Fraction<-rep("100um",nrow(hetero.100))

f="53LTR"
hetero.53 <- heterotroph_test(physeq = moth_good,oligo.dat = oligo.dat,f = "53LTR",prune = 500)
hetero.53$Fraction<-rep("53um",nrow(hetero.53))

f="3NA"
hetero.3 <- heterotroph_test(physeq = moth_good,oligo.dat = oligo.dat,f = "3NA",prune = 500)
hetero.3$Fraction<-rep("3um",nrow(hetero.3))

f="22NA"
hetero.22NA <- heterotroph_test(physeq = moth_good,oligo.dat = oligo.dat,f = "22NA",prune = 500)
hetero.22NA$Fraction<-rep(".22um",nrow(hetero.22NA))


f="CNA"
hetero.CNA <- heterotroph_test(physeq = moth_good,oligo.dat = oligo.dat,f = "CNA",prune = 500)

# rbind them all
allhetero<-rbind(hetero.100,hetero.53,hetero.3,hetero.22NA)
names(allhetero)[4]<-"Phylum"
names(allhetero)[5]<-"Class"

allhetero.summary<-count(allhetero,vars=c("Class","Fraction","Type"))
allhetero.summary$Fraction<-factor(allhetero.summary$Fraction,levels= c("100um","53um","3um",".22um"))

ggplot(allhetero.summary,aes(x=Class,y=freq,group=Type,fill=Type))+geom_bar(stat="identity")+facet_wrap(~Fraction,ncol=1) + ylab("Number of correlated OTUs \n") + xlab("")+theme(axis.text.x = element_text(angle=45,vjust=1, 
                                       hjust = 1, size=10,face="bold"))


```



# Alpha Diversity
```{r alpha diversity}

alphadiv <- read.table("other/alphadiv.txt",header=TRUE)

alphadiv <- order_dates(alphadiv)

alphadiv$Station <- factor(alphadiv$Station,levels = c("WE2","WE4","WE12"))
alphadiv$variable <- factor(alphadiv$variable,levels=c("Richness","Evenness"))
alphadiv$value[58:114] <- alphadiv$value[58:114]/alphadiv$value[1:57]


theme_set(theme_classic())

pall <- ggplot(alphadiv,aes(x=Date,y=value,color=Station,group=Station,shape=Station)) +
        geom_point(size=2) + 
        geom_line(size=.8) +
        facet_wrap(~variable,ncol=1,scales="free") +
        ylab("") +
        xlab("") +
        ggtitle("") +
        scale_color_manual(values=c("#E96446","#302F3D","#87CEFA")) +
        theme(axis.text.x = element_text(angle=45,vjust=1, 
                                       hjust = 1, size=10,face="bold"),
              plot.margin = unit(c(0,0,1,0), "lines"),
              strip.background=element_rect(color="white"),
              strip.text=element_text(face="bold",size = 14),
              panel.margin = unit(1, "lines"))

           

p1 <- ggplot(richness,
             aes(x=Date,y=value,color=Station,group=Station,shape=Station)) +
      geom_point() +
      geom_line() +
      ylab("Richness") +
      xlab("") +
      ggtitle("") +
  
scale_color_manual(values=c("#E96446","#302F3D","#87CEFA"))+
theme(axis.text.x = element_text(angle=45,vjust=1, hjust = 1, size=10, face="bold"))
           


p2 <- ggplot(evenness,aes(x=Date,y=value,color=Station,group=Station,shape=Station))+
      geom_point()+geom_line()+
      ylab("ISI")+xlab("")+
      ggtitle("")+
      scale_color_manual(values=c("#E96446","#302F3D","#87CEFA"))+
      theme(axis.text.x = element_text(angle=45,vjust=1, hjust = 1, size=10, face="bold"))

grid.arrange(m2, arrangeGrob(pall, ncol=1), ncol=2, widths=c(1.2,1))


# linear models

unmelt<-data.frame(alphadiv[1:57,-3],alphadiv[58:114,4])
names(unmelt)[3:4]<-c("Richness","Evenness")

alpha.join<-merge(sample_data(cna),unmelt,by.x=c("Date","Station"))

logphyco.rich.fit <- lm(Phycocyanin~Richness,data=alpha.join)
logphyco.even.fit <- lm(LogPhyco~Evenness,data=alpha.join)

summary(logphyco.rich.fit)
summary(logphyco.even.fit)

```



# Ordinations

## Unconstrained
```{r}
CNA <- subset_samples(moth_good,Fraction=="CNA")

CNA.scale <- scale_reads(CNA,min(sample_sums(CNA)))

sample_data(CNA.scale)$Station <- factor(sample_data(CNA.scale)$Station,
                                         levels=c("WE2","WE4","WE12"))

sample_data(CNA.scale)$Month <- factor(sample_data(CNA.scale)$Month,
                                       levels=c("May","June","July","August",
                                                "September","October","November"))

cna.scale.bray= ordinate(CNA.scale, method="PCoA", "bray")

theme_set(theme_bw())
plot_ordination(physeq = CNA.scale,
                              ordination = cna.scale.bray,
                              color = "Month",
                              shape = "Station",
                              axes = c(1,2)) + 
              geom_point(aes(colour = Month),alpha=.7, size = 4) +
              geom_point(colour="grey90", size = 1.5) + 
              scale_color_manual(values=c("#a65628","red","#ffae19",
                                          "#4daf4a","#1919ff",
                                          "darkorchid3","magenta")) +
              theme(plot.margin = unit(c(2,2,2,2), "lines"))


st2and12<-subset_samples(CNA,Station!="WE4")
doadonis(physeq = st2and12,category = "Station")

s<-data.frame(sample_data(CNA))

s$Station[s$Station=="WE12"]<-"WE2"
s<-sample_data(s)
newstations<-CNA
sample_data(newstations)<-s

doadonis(newstations,"Station")

nonbloommonths<-subset_samples(physeq = CNA, Month %in%
                                 c("May","June","October"))
doadonis(nonbloommonths,category = "Station")

bloommonths<-subset_samples(physeq=CNA, Month %in%
                              c("July","August","September"))

s<-data.frame(sample_data(bloommonths))

s$Station[s$Station=="WE12"]<-"WE2"
s<-sample_data(s)
newstations<-bloommonths
sample_data(newstations)<-s
doadonis(newstations,category = "Station")

st2and12<-subset_samples(bloommonths,Station!="WE4")
doadonis(physeq = st2and12,category = "Station")
```


## Constrained 

```{r}

cna <- subset_samples(moth_good,Fraction=="CNA")
# cna <-subset_samples(cna,Station!="WE4")
cna.scale<-scale_reads(cna,n = 10000)
cna.scale<-prune_samples(sample_sums(cna.scale)>0,cna.scale)


cna.scale.sub<-subset_samples(cna.scale,!is.na(LogPhyco) & !is.na(SRP)&
                                !is.na(DOC) & !is.na(pH)&!is.na(ParMC)&!is.na(H2O2))

bdist<-phyloseq::distance(physeq = cna.scale.sub,method = "bray")

sample_data(cna.scale.sub)$Month <- factor(sample_data(cna.scale.sub)$Month,
                                           levels=c("May","June","July","August",
                                                    "September","October","November"))
                                    


# CAP
cap.ord<-ordinate(physeq = cna.scale.sub,method = "CAP",
                  distance = bdist,
                  formula = ~ParMC+Nitrate+SRP+LogPhyco+Ammonia+DOC+pH+H2O2)


cap.plot <- plot_ordination(cna.scale.sub,cap.ord,color="Month",axes = c(1,2)) + 
            aes(shape = Station) +
            geom_point(aes(colour = Month),alpha=.2, size = 4) +
            geom_point(colour="grey90", size = 1.5) + 
            scale_color_manual(values=c("#a65628","red","#ffae19",
                                        "#4daf4a","#1919ff",
                                        "darkorchid3","magenta"))


# Now add the environmental variables as arrows
arrowmat = vegan::scores(cap.ord, display = "bp")

# Add labels, make a data.frame
arrowdf <- data.frame(labels = rownames(arrowmat), arrowmat)

# Define the arrow aesthetic mapping
arrow_map = aes(xend = CAP1, yend = CAP2, x = 0, y = 0, 
                shape = NULL, color = NULL, 
                label = labels)

label_map = aes(x = 1.3 * CAP1, y = 1.3 * CAP2, shape = NULL, color = NULL, 
    label = labels)

arrowhead = arrow(length = unit(0.02, "npc"))

# Make a new graphic
p1 = cap.plot + 
     geom_segment(arrow_map, size = .5, 
                             data = arrowdf, color = "gray", 
                             arrow = arrowhead) + 
    geom_text(label_map, size = 4,face="bold", 
                data = arrowdf,show_guide=FALSE)


anova(cap.ord)


multiplot(months.plot,p1,cols = 2)


```

# mantel tests
```{r}


bdist<-phyloseq::distance(cna.scale,"bray")
timedist<-dist(sample_data(cna.scale)$Days)
phycodist<-dist(sample_data(cna.scale)$Phycocyanin)
chladist<-dist(data.frame(sample_data(cna.scale))$Chla)
logphycodist<-dist(sample_data(cna.scale)$LogPhyco)
logchladist<-dist(data.frame(sample_data(cna.scale))$LogChla)
logparmcdist<-dist(data.frame(sample_data(cna.scale))$LogParMC)

# Simple mantel
mantel(bdist~timedist)
mantel(bdist,timedist,method = "spearman")

# Partial mantel
library(ecodist)

# Phycocyanin
phyco.na<-which(is.na(phycodist))
phycodist.na<-phycodist[-phyco.na]
timedist.na<-timedist[-phyco.na]
bdist.na<-bdist[-phyco.na]
# Effect of time, removing effect of phycocyanin
mantel(formula = bdist.na~timedist.na+phycodist.na)
# Effect of phycocyanin, removing effect of time
mantel(formula = bdist.na~phycodist.na+timedist.na)

# Chla
chla.na<-which(is.na(chladist))
chladist.na<-phycodist[-chla.na]
timedist.na<-timedist[-chla.na]
bdist.na<-bdist[-chla.na]
# Effect of time, removing effect of phycocyanin
mantel(formula = bdist.na~timedist.na+chladist.na)
# Effect of chla, removing effect of time
mantel(formula = bdist.na~chladist.na+timedist.na)


# log phyco
logphyco.na<-which(is.na(logphycodist))
logphycodist.na<-logphycodist[-logphyco.na]
timedist.na<-timedist[-logphyco.na]
bdist.na<-bdist[-logphyco.na]
mantel(bdist.na~logphycodist.na)
mantel(formula = bdist.na~timedist.na+logphycodist.na)
mantel(formula = bdist.na~logphycodist.na+timedist.na)


# log chla
logchla.na<-which(is.na(logchladist))
logchladist.na<-logchladist[-logchla.na]
timedist.na<-timedist[-logchla.na]
bdist.na<-bdist[-logchla.na]
mantel(formula = bdist.na~timedist.na+logchladist.na)
mantel(formula = bdist.na~logchladist.na+timedist.na)

# log parmc 
logparmc.na<-which(is.na(logparmcdist))
logparmcdist.na<-logparmcdist[-logparmc.na]
timedist.na<-timedist[-logparmc.na]
bdist.na<-bdist[-logparmc.na]
mantel(formula = bdist.na~timedist.na+logparmcdist.na)
mantel(formula = bdist.na~logparmcdist.na+timedist.na)

# Correlogram
mgram.1<-mgram(bdist,timedist,breaks=seq(0,170,by=7))
plot(mgram.1,axes=FALSE,xlab="Time (days)")
axis(side=1, at=seq(0,170,by=14))
axis(side=2, at=seq(-.25, .6, by=.05))
abline(h=0,col="blue")
```

