---
title: "HABs 16s August 2015"
output:
  html_document:
    toc: true
    theme: united
---

The main goals for this week were:


1) Examine patterns of diversity within the cyanobacteria
    - How many OTUs of each genus are there?
    - Shifts in OTUs over time - is MC different than other cyanos?
    - How do shifts in oligotypes compare to OTUs?

2) look at non-abundance weighted measures of dissimilarity to see if there are real turnover events in the community throughout the season

3) Explore Erens MED analysis and rerun some ordinations to see if they are similar overall 


```{r load libraries, warning = FALSE, message = FALSE, echo=FALSE}
#Load libraries
library(phyloseq)
library(Biostrings)
library(ggplot2)
library(vegan)
library(plyr)
library(scales)
library(grid)
library(reshape2)
library(gridExtra)
library(RColorBrewer)
source("habs_functions.R")
```



```{r mothur import, echo = FALSE}

# Import mothur files and sample metadata
sharedfile = "mothur/allhabs.shared"
taxfile = "mothur/allhabs.taxonomy"
mapfile = "other/best_metadata.csv"
 
mothurdata = import_mothur(mothur_shared_file = sharedfile, 
	mothur_constaxonomy_file = taxfile)

# Add the OTU number as a column in the taxonomy file
tax_table(mothurdata) <- cbind(tax_table(mothurdata), 
	row.names(tax_table(mothurdata)))

# Rename the taxonomy columns
colnames(tax_table(mothurdata)) <- c("Kingdom", "Phylum", "Class", "Order", 
	"Family", "Genus", "Species")
  
# Import sample metadata
map <- read.csv(mapfile)
map <- sample_data(map)
rownames(map) <- map$SampleID
  
# Merge mothurdata object with sample metadata
moth_merge = merge_phyloseq(mothurdata, map)

# Filter out non-samples (i.e. water,mock) and samples from intensive cruises.
# Also prune out taxa which were only present in removed samples
moth_good <- subset_samples(moth_merge,Type == "sample")
moth_good <- prune_taxa(taxa_sums(moth_good) > 0, moth_good)

# Filter out non-bacteria, chloroplasts and mitochondria
moth_good <- subset_taxa(moth_good, Kingdom == "Bacteria")
moth_good <- subset_taxa(moth_good, Class != "Chloroplast")
moth_good <- subset_taxa(moth_good, Family != "mitochondria")

```

```{r representative sequences,echo=FALSE,include=FALSE}

# Import representative sequences from mothur
ref.seq <- readDNAStringSet("mothur/mothur_repseq.fasta")

# Split on tab to seperate sequence header and OTU number
ref1 <- data.frame(do.call('rbind',
                           strsplit(as.character(names(ref.seq)),'\t',
                                    fixed=TRUE)))
# Split on | to seperate OTU number
ref2 <- data.frame(do.call('rbind',
                           strsplit(as.character(ref1$X2),"|",
                                    fixed=TRUE)))

# Make the OTU names the rownames for the reference sequences
names(ref.seq)<-ref2$X1

# Turn into refseq object
ref.seq <- refseq(ref.seq)


```


# Stacked barplot

Time series of phylum community composition across filter fractions and 
sampling stations


```{r composition barplot, echo=FALSE, fig.height = 10, fig.width = 14}

# Set phylum colors
phylum.colors <- c("#CBD588", "#5F7FC7", "orange", "#DA5724", "#89C5DA",
	"#508578", "#CD9BCD", "#74D944", "#D7C1B1", "#AD6F3B", "#673770","#D14285",
	"#689030", "#6DDE88", "#652926", "#7FDCC0", "#C84248", "#8569D5", "#5E738F",
	"#D1A33D", "#8A7C64", "#599861"
)


# Transform to long format and prune out phyla below 2% in each sample
moth.long <- transform_and_melt(physeq = moth_good,
	taxrank = "Phylum",
	prune = .05)

# Plot 
make_tax_barplot(
  df = moth.long, 
  x = "Date", 
  y = "Abundance",
  tax = "Phylum",
  facet = "Fraction~Station",
  title = "",
  colors = phylum.colors,
  xlab = "",
  ylab = "Relative Abundance\n",
  outline = "black",
  relative = TRUE,
  guide = TRUE
)


```

# Ordinations

## Unconstrained

### Bray-Curtis

```{r}
theme_set(theme_bw())

myord(
	physeq = moth_good, 
	fraction = "CNA", 
	method = "PCoA", 
	distance = "bray", 
	title = "Full community Bray-Curtis PCoA"
)
```

```{r}

myord(
	physeq = moth_good, 
	fraction = "100LTR", 
	method = "PCoA", 
	distance = "bray", 
	title = "100um Bray-Curtis PCoA"
)

```


```{r}
# 53-100um fraction PCoA with Sorenson distance
myord(
	physeq = moth_good, 
	fraction = "53LTR", 
	method = "PCoA", 
	distance = "bray", 
	title = "53um Bray-Curtis PCoA"
)


```

```{r}
# 3-53um fraction PCoA with Sorenson distance
myord(
	physeq = moth_good, 
	fraction = "3NA", 
	method = "PCoA", 
	distance = "bray", 
	title = "3um Bray-Curtis PCoA"
)


```

```{r}
# 0.22um-3um fraction PCoA with Bray-Curtis distance
myord(
	physeq = moth_good, 
	fraction = "22NA", 
	method = "PCoA", 
	distance = "bray", 
	title = ".22um Bray-Curtis PCoA"
)

```

### Sorenson 

```{r}
# Full community fraction PCoA with Sorenson distance
myord(
	physeq = moth_good, 
	fraction = "CNA", 
	method = "PCoA", 
	distance = "sor", 
	title = "Full community Sorenson PCoA"
)
```

```{r}
# >100um fraction PCoA with Sorenson distance
myord(
	physeq = moth_good, 
	fraction = "100LTR", 
	method = "PCoA", 
	distance = "sor", 
	title = "100um Sorenson PCoA"
)

```


```{r}
# 53-100um fraction PCoA with Sorenson distance
myord(
	physeq = moth_good, 
	fraction = "53LTR", 
	method = "PCoA", 
	distance = "sor", 
	title = "53um Sorenson PCoA"
)

```


```{r}
# 3-53um fraction PCoA with Sorenson distance
myord(
	physeq = moth_good, 
	fraction = "3NA", 
	method = "PCoA", 
	distance = "sor", 
	title = "3um Sorenson PCoA"
)

```


```{r}
# Free living fraction PCoA with Sorenson distance
myord(
	physeq = moth_good, 
	fraction = "22NA", 
	method = "PCoA", 
	distance = "sor", 
	title = ".22um Sorenson PCoA"
)
```

## Constrained 

```{r echo = FALSE}
# Subset to just full community
cna.scale <- scale_sub(physeq = moth_good,fraction = "CNA")

# Remove data points with missing metadata
cna.scale.sub <-
  subset_samples(
	cna.scale,!is.na(LogPhyco) & 
	  !is.na(SRP) &
	  !is.na(DOC) &
	  !is.na(pH) & 
	  !is.na(ParMC) & 
	  !is.na(H2O2)
  )

bdist <- phyloseq::distance(physeq = cna.scale.sub, method = "bray")

							
# CAP ordinate
cap.ord <- ordinate(
	physeq = cna.scale.sub, 
	method = "CAP",
	distance = bdist,
	formula = ~ ParMC + Nitrate + SRP + LogPhyco + Ammonia + DOC + pH + H2O2
)

# CAP plot
cap.plot <- plot_ordination(
	cna.scale.sub, 
	cap.ord, 
	color = "Month", 
	axes = c(1,2)
	) + 
	aes(shape = Station) + 
	geom_point(aes(colour = Month), 
		alpha= 0.2, 
		size = 4) + 
	geom_point(colour = "grey90", 
		size = 1.5) + 
	scale_color_manual(values = c("#a65628", "red", "#ffae19","#4daf4a", 
		"#1919ff", "darkorchid3", "magenta"))


# Now add the environmental variables as arrows
arrowmat <- vegan::scores(cap.ord, display = "bp")

# Add labels, make a data.frame
arrowdf <- data.frame(labels = rownames(arrowmat), arrowmat)

# Define the arrow aesthetic mapping
arrow_map <- aes(xend = CAP1, 
	yend = CAP2, 
	x = 0, 
	y = 0, 
	shape = NULL, 
	color = NULL, 
	label = labels)

label_map <- aes(x = 1.3 * CAP1, 
	y = 1.3 * CAP2, 
	shape = NULL, 
	color = NULL, 
	label = labels)

arrowhead = arrow(length = unit(0.02, "npc"))

# Make a new graphic
cap.plot + 
geom_segment(arrow_map, 
	size = .5, 
	data = arrowdf, 
	color = "gray", 
	arrow = arrowhead
	) + 
geom_text(label_map, 
	size = 4, face = "bold", 
	data = arrowdf, 
	show_guide = FALSE)

```

Do a permutational ANOVA on constrained axes used in ordination

```{r}

anova(cap.ord)

```

# Cyanobacteria analysis

## Genus diversity

```{r echo=FALSE}
cyanos <- subset_taxa(physeq = moth_good, Phylum == "Cyanobacteria")

# How many reads of different genera are there? 
cyano_genus <- tax_glom(cyanos, taxrank = "Genus")
cyano_sums <- taxa_sums(cyano_genus)
names(cyano_sums) <- tax_table(cyano_genus)[,"Genus"]
```

How many reads of different genera are there?
```{r}
print(cyano_sums)
```


How many OTUs are there of each genus?
```{r}
gen <- tax_table(cyanos)[,6]
table(gen)
```

```{r echo=FALSE}
# Transform sample counts
all.proportions <- transform_sample_counts(moth_good, function(x) {x/sum(x)})

anab <- subset_taxa(physeq = all.proportions, Genus == "Anabaena")
pseu <- subset_taxa(physeq = all.proportions, Genus == "Pseudanabaena")
syne <- subset_taxa(physeq = all.proportions, Genus == "Synechococcus")
micr <- subset_taxa(physeq = all.proportions, Genus == "Microcystis")
OTU49 <- subset_taxa(physeq = all.proportions, Species == "Otu00049")
OTU367 <- subset_taxa(physeq = all.proportions, Species == "Otu00367")

```

## OTU vs Oligotype barplots

```{r echo = FALSE, fig.height = 10, fig.width = 14}
sampdat <- data.frame(
  sample_data(moth_good)[ , c("Date", "SampleID", "Station", "Fraction") ] 
)
sampdat$Sums <- sample_sums(moth_good)

# Anabaena OTUs
anab.long <- cyano_melt(anab)

make_tax_barplot(
  df = anab.long, 
  x = "Date", 
  y = "Abundance",
  tax = "Species",
  facet = "Fraction~Station",
  title = "Anabaena OTUs",
  colors = phylum.colors,
  xlab = "",
  ylab = "Relative Abundance\n",
  outline = "black",
  relative = FALSE,
  guide = TRUE
)


# Anabaena oligotypes
ana.oligo <- read.csv("~/chabs/miseq_may2015/oligotype/anabaena/anabaena_oligotyping-sc3-s1-a0.0-A0-M50/MATRIX-COUNT.txt", sep = "\t")

ana.melt <- oligo_prep(ana.oligo, sampdat)
oligoplot(ana.melt, "Anabaena Oligotypes")

```

```{r echo=FALSE, fig.height = 10, fig.width = 14}


pseu.long <- cyano_melt(pseu)

make_tax_barplot(
  df = pseu.long, 
  x = "Date", 
  y = "Abundance",
  tax = "Species",
  facet = "Fraction~Station",
  title = "Pseudanabaena",
  colors = phylum.colors,
  xlab = "",
  ylab = "Relative Abundance\n",
  outline = "black",
  relative = FALSE,
  guide = TRUE
)

# Anabaena oligotypes
pseu.oligo <- read.csv("~/chabs/miseq_may2015/oligotype/pseudanabaena/pseudanabaena_oligotyping-sc2-s1-a0.0-A0-M50/MATRIX-COUNT.txt", sep = "\t")

pseu.melt <- oligo_prep(pseu.oligo, sampdat)
oligoplot(pseu.melt, "Pseudanabaena Oligotypes")

```


```{r echo=FALSE, fig.height = 10, fig.width = 14}
syne.long <- cyano_melt(syne)

make_tax_barplot(
  df = syne.long, 
  x = "Date", 
  y = "Abundance",
  tax = "Species",
  facet = "Fraction~Station",
  title = "Synechococcus",
  colors = phylum.colors,
  xlab = "",
  ylab = "Relative Abundance\n",
  outline = "black",
  relative = FALSE,
  guide = TRUE
)


# synechococcus 
syne.oligo <- read.csv("~/chabs/miseq_may2015/oligotype/synechococcus80/synecohococcus_oligotype-m0.10-A0-M0-d4/MATRIX-COUNT.txt", sep = "\t")

syne.melt <- oligo_prep(syne.oligo, sampdat)
oligoplot(syne.melt, "Synechococcus Oligotypes (cutoff 80)")


syne.oligo90 <- read.csv("~/chabs/miseq_may2015/oligotype/synechococcus90/synechococcus90_oligotype-m0.10-A0-M0-d4/MATRIX-COUNT.txt", sep = "\t")

syne.melt90 <- oligo_prep(syne.oligo90, sampdat)
oligoplot(syne.melt, "Synechococcus Oligotypes (cutoff 90)")


```

```{r echo = FALSE, fig.height = 10, fig.width = 14}

mc.long <- cyano_melt(micr)

make_tax_barplot(
  df = mc.long, 
  x = "Date", 
  y = "Abundance",
  tax = "Species",
  facet = "Fraction~Station",
  title = "Microcystis",
  colors = phylum.colors,
  xlab = "",
  ylab = "Relative Abundance\n",
  outline = "black",
  relative = FALSE,
  guide = TRUE
)

# Microcystis oligotypes
mc.oligo <- read.csv("~/chabs/miseq_may2015/oligotype/microcystis/allhabs_mc_oligo-c2-s1-a0.0-A0-M50/MATRIX-COUNT.txt", sep = "\t")

mc.melt <- oligo_prep(mc.oligo, sampdat)
oligoplot(mc.melt, "Microcystis")

```


```{r echo=FALSE, fig.height = 10, fig.width = 14}

otu49.long <- cyano_melt(OTU49)

make_tax_barplot(
  df = otu49.long, 
  x = "Date", 
  y = "Abundance",
  tax = "Species",
  facet = "Fraction~Station",
  title = "OTU49",
  colors = phylum.colors,
  xlab = "",
  ylab = "Relative Abundance\n",
  outline = "black",
  relative = FALSE,
  guide = TRUE
)

OTU367.long <- cyano_melt(OTU367)

make_tax_barplot(
  df = OTU367.long, 
  x = "Date", 
  y = "Abundance",
  tax = "Species",
  facet = "Fraction~Station",
  title = "OTU367",
  colors = phylum.colors,
  xlab = "",
  ylab = "Relative Abundance\n",
  outline = "black",
  relative = FALSE,
  guide = TRUE
)

```

# Conclusions
```{r}

```

1) Examine patterns of diversity within the cyanobacteria
    - How many OTUs of each genus are there?
    - Shifts in OTUs over time - is MC different than other cyanos?
    - How do shifts in oligotypes compare?

OTU analysis:
There are up to 10 OTUs for different genera, but there is usually only a couple abundant OTUs (others are spurious??). Anabaena has 2, Synechococcus has 4, Microcystis and pseudanabaena have only 1 dominant OTU. 

There are no obvious patterns of shifts in OTU populations over time, except for synechococcus. The orange OTU only appears in the first part of the season and the red appears in the second. Also, it seems like synechococcus blooms slightly ahead of microcystis. 

Both OTU 49 and 367 had no matches in NCBI to known things


Oligotype analysis: 
Synechococcus gets the award for most interesting organism this week!!!
Already at the OTU level we can see more diversity in syn than any other cyano genera, but after oligotyping there is an astonishing amount of diversity. Is this real?? The BLAST results from each of the oligotypes show mixed results; Some were listed as synechococcus others as unclassified cyanobacteria. I went back to my taxonomy file and got rid of all sequences with a bootstrap value below 90 for synechococcus. The entropy analysis still reveals almost the same amount of diversity





2) look at non-abundance weighted measures of dissimilarity to see if there are real turnover events in the community throughout the season

It looks like most changes in community composition are due to changes in relative abundance rather than species turnover

3) Explore Erens MED analysis and rerun some ordinations to see if they are similar overall 

A few problems with MED

1) MED paper reccomends running it with -M parameter N/10,000 or larger (i.e about 1500 for my dataset) Whats the point of this pipeline if youre going to throw away all of your rare (not even that rare) OTUs? 

2) Entropy is biased based on library size. Since these samples have very different sequencing depths, I would need to normalize before running it to get accurate results

Ultimately MED is just an algorithmic improvement to Oligotyping, not a way to avoid OTU clustering with whole community data

