---
title: "16s Pipeline Comparison"
output:
  html_document:
    toc: true
    theme: united
---


In this document I use the Lake Erie HABs dataset to illustrate differences in the mothur and uparse pipelines. For all pipelines we used used the silva database v119 as our taxonomy reference. 

## Load required libraries
```{r,warning=FALSE,message=FALSE}
#Load libraries
library(phyloseq)
packageVersion("phyloseq")
library(ggplot2)
library(ape)
library(vegan)
library(plyr)
library(scales)
library(grid)
library(reshape2)
theme_set(theme_bw())

# Set working directory
setwd("~/habs/genomics/r/data/")
```

```{r,echo=FALSE}
## Functions 

scale_reads <- function(physeq,n){
  physeq.scale <- transform_sample_counts(physeq, function(x) {(n*x/sum(x))})
  otu_table(physeq.scale) = floor(otu_table(physeq.scale))
  physeq.scale = prune_taxa(taxa_sums(physeq.scale)>0,physeq.scale)
  return(physeq.scale)
}

order_dates <- function(df){
  df$Date <- factor(df$Date, levels = c("5/27/14","6/10/14","6/16/14","6/30/14","7/8/14","7/14/14","7/21/14","7/29/14","8/4/14","8/11/14","8/18/14","8/25/14","9/2/14","9/8/14","9/15/14","9/23/14","9/29/14","10/6/14","10/15/14","10/20/14","10/27/14","11/3/14"))
  return(df)
}

### Data import

mothur_import_and_clean<-function(sharedfile,taxfile,mapfile){
  
  mothurdata = import_mothur(mothur_shared_file = sharedfile,
                             mothur_constaxonomy_file = taxfile )

  # We need to change the taxonomy names
  tax_table(mothurdata)<-cbind(tax_table(mothurdata),row.names(tax_table(mothurdata)))
  colnames(tax_table(mothurdata)) <- c("Kingdom","Phylum","Class","Order","Family","Genus","Species")
  
  #import mapfile
  map = import_qiime_sample_data(mapfile)
  colnames(otu_table(mothurdata))<-map$SampleID
  moth_merge = merge_phyloseq(mothurdata,map)

  #Filter out shit 
  moth_good <- subset_samples(moth_merge,Type=="sample")
  moth_good <- subset_samples(moth_good,Intensive=="n")
  moth_good <- subset_taxa(moth_good, Kingdom == "Bacteria")
  moth_good <- subset_taxa(moth_good, Class != "Chloroplast")
  moth_good <- subset_taxa(moth_good, Family!="mitochondria")
  moth_good <- prune_taxa(taxa_sums(moth_good)>0,moth_good)
  return(moth_good)
}

uparse_import_and_clean <- function(otufile,taxfile,mapfile){
  
  # Uparse import
  otu<-read.csv(otufile,sep="\t")
  tax<-read.csv(taxfile,sep="\t")
  rownames(tax)<-rownames(otu)
  map<-read.csv(mapfile,sep="\t")
  rownames(map)<-map$SampleID
  phyOTU<-otu_table(otu,taxa_are_rows=TRUE)
  tax<-as.matrix(tax)
  phyTax<-(tax_table(tax))
  phySam<-sample_data(map)
  uparse.physeq<-phyloseq(phyOTU,phyTax,phySam)
  
  # Filter out bad shit
  uparse.good<-subset_taxa(uparse.physeq,
                           Class!="Chloroplast" & 
                             Kingdom =="Bacteria" & 
                             Family!="mitochondria")
  uparse.good <- subset_samples(uparse.good,Type=="sample" & Intensive=="n")
  uparse.good <- prune_taxa(taxa_sums(uparse.good)>0,uparse.good)
  
  return(uparse.good)
  
}



make_barplot<-function(physeq,title){
  # Agglomerate all otu's by phylum level
  physeq_phylum<-tax_glom(physeq,taxrank = "Phylum")

  # Create a new phyloseq object which for each sample only includes phyla with a relative abundance above 1 % 
  physeq_phylum.99 = transform_sample_counts(physeq_phylum, 
                                              function(x){x/sum(x)})
  otu_table(physeq_phylum.99)[otu_table(physeq_phylum.99)<.01] <- 0 
  physeq_phylum.99 = prune_taxa(taxa_sums(physeq_phylum.99)>0,
                                 physeq_phylum.99)

  # Melt into long format and sort by samplenum and phylum
  physeq_phylum.long<-psmelt(physeq_phylum.99)
  physeq_phylum.long<-arrange(physeq_phylum.long,Samplenum,Phylum)

  # Set phylum plotting colors
  phylum.colors <- c(Acidobacteria="darkorchid3",Actinobacteria="#0099ff",Armatimonadetes="orange", Bacteroidetes="#0c2b57",Candidate_division_SR1 = "#AD6F3B",Candidate_division_OD1="purple",Chlorobi="cyan", Chloroflexi="magenta",Cyanobacteria="#7dc107",Firmicutes="yellow",Gemmatimonadetes="maroon",Nitrospirae="royalblue",Planctomycetes="darkgreen",Proteobacteria="#E12C4D", TM6="white",unclassified="grey", Verrucomicrobia="black")

  # Reorder factor levels for dates 
  physeq_phylum.long<-order_dates(physeq_phylum.long)

  # Reorder factor levels for Fraction and Station. Change Fraction factor names
 physeq_phylum.long$Fraction <- factor(physeq_phylum.long$Fraction,
                                       levels=c("CNA","100L","22NA"))
  levels(physeq_phylum.long$Fraction) <- c("Full Community", 
                                           "Particle/Algal", "Free Living")
  physeq_phylum.long$Station <-factor(physeq_phylum.long$Station,
                                      levels=c("WE2","WE12","WE4"))


  # GGPLOT it
  ggplot(physeq_phylum.long,aes(x=Date,y=Abundance,fill=Phylum)) +
        facet_grid(Fraction~Station,scales="free",space="free")+
        geom_bar(stat="identity",position="fill")+ 
   scale_y_continuous(labels = percent_format())+ 
  scale_fill_manual(values = phylum.colors,name="Phylum") + 
    theme(axis.title.x = element_text(size=16,face="bold"),
          axis.text.x = element_text(angle=50, colour = "black", vjust=1, hjust = 1, size=6.5),
          axis.text.y = element_text(colour = "black", size=9),
          axis.title.y = element_text(face="bold", size=16),
          plot.title = element_text(size = 18),
          legend.title = element_text(size=14),
          legend.text = element_text(size = 13),
          legend.position="right")+
    guides(fill = guide_legend(reverse= TRUE,keywidth=1,keyheight=1))+
    xlab("Sampling Date")+
    ylab("Relative Abundance (taxa>.1%)")+
    ggtitle(title)
        
}


doadonis <- function(physeq, category) {
  physeq.scale <- scale_reads(physeq,min(sample_sums(physeq)))
  bdist<-phyloseq::distance(physeq.scale,"bray")
  col = as(sample_data(physeq),"data.frame")[,category]
  adonis.bdist=adonis(bdist~col)
  print(adonis.bdist) 
 

 # Homogeneity of dispersion test
  betatax=betadisper(bdist,col)
  p=permutest(betatax)
  print("The homogeneity of dispersion:") 
  print(p$tab)
}  

```


## Mothur
### SOP (cutoff=80)
First, we will import the shared file and taxonomy file outputs from Mothur ran according to the the SOP. Specifically, this used an 80% cutoff to assign taxonomy to sequences and clustered OTUs by first dividing sequences at the Order level (taxrank=4) and clustering within the Oder. A taxonomy is assigned to an OTU if at least 51% of the sequences in the OTU were assigned the same taxonomy. 

```{r}

### Data import

mothursop<-mothur_import_and_clean(sharedfile = "newsilva.shared",taxfile = "newsilva.taxonomy",mapfile = "metadata_no1.txt")

mothursop
```



In our mothusop object we have 12,126 OTUs and 183 samples. In this object I have already filtered out samples that did not amplify, blanks, mock communities, and samples from the intensive cruises. I have also removed non bacterial taxa like archaea, mitochondria, and chloroplasts



Okay, first lets plot a histogram of sample read counts
```{r,echo=FALSE,warning=FALSE,message=FALSE}
# Histogram of sample read counts
ggplot(data.frame(sum=sample_sums(mothursop)),aes(sum)) + geom_histogram(colour="white",fill="indianred")+ggtitle("Sample read counts (mothur SOP)")+xlab("total sequences")

```

A lot of our samples were trashed when removing chloroplasts since we had a lot of reads from eukaryotic algae in our 100um samples. 

Now lets look at a stacked barplot of community composition
```{r,fig.height=8,fig.width=12}
make_barplot(physeq = mothursop,title = "Mothur 80 (new silva)")
```


In this barplot we see pretty low diversity. There are only 4 dominant phyla; actinobacteria, protebacteria, cyanobacteria, and Bacteroidetes. There are also a large number of unknowns in ever filter fraction. 

I also want to compare an OTU-based analysis across methods. For this I will run an adonis test on the bray-curtis distance between samples and the date the samples were taken
```{r}
# Subset to just whole community fraction 
CNA.moth80 <- subset_samples(mothursop,Fraction=="CNA")
CNA.moth80

# Run adonis
doadonis(physeq = CNA.moth80,category = "Date")



````

### Cutoff = 60
For this data we followed the same procedure as the SOP except we ran the classify.seqs command with cutoff = 60. 

```{r}
mothur60<-mothur_import_and_clean(sharedfile = "moth60.shared",taxfile = "moth60.taxonomy",mapfile = "metadata.txt")

mothur60

```


Now we have 11,621 OTUs accross our 183 samples. I don't really understand why this produces MORE OTUs, but it must have to do with the cluster.split command which uses the taxonomy file as a parameter. This is supposed to reduce ram usage, but i think it is biasing the otu clustering results. 


The distribution of sample read counts looks the same
```{r,echo=FALSE,warning=FALSE}
# Histogram of sample read counts
ggplot(data.frame(sum=sample_sums(mothur60)),aes(sum)) + geom_histogram(colour="white",fill="indianred")+ggtitle("Sample read counts (mothur 60)")+xlab("total sequences")

```

We can see in this stacked barplot that we know have more firmicutes,planctomycetes,verrucomicrobia and armatimonadetes classified in our communities now. 
```{r,fig.height=8,fig.width=12}
make_barplot(physeq = mothur60,title = "Mothur 60")
```

```{r}
# Subset to just whole community fraction 
CNA.moth60 <- subset_samples(mothur60,Fraction=="CNA")
CNA.moth60

# Run adonis
doadonis(physeq = CNA.moth60,category = "Date")



````
### Cutoff = 51
For this data we followed the same procedure as the SOP except we ran the classify.seqs command with cutoff = 51. 

```{r,echo=FALSE}
### Data import

mothur51<-mothur_import_and_clean(sharedfile = "moth51.shared",taxfile = "moth51.taxonomy",mapfile = "metadata.txt")

mothur51
```


The distribution of sample read counts looks the same
```{r,echo=FALSE,warning=FALSE}
# Histogram of sample read counts
ggplot(data.frame(sum=sample_sums(mothur51)),aes(sum)) + geom_histogram(colour="white",fill="indianred")+ggtitle("Sample read counts (mothur 51)")+xlab("total sequences")

```

Now we have even more firmicutes with a few more planctomycetes and verrucos
```{r,fig.height=8,fig.width=12}
make_barplot(physeq = mothur51,title = "Mothur 51")
```

```{r}
# Subset to just whole community fraction 
CNA.moth51 <- subset_samples(mothur51,Fraction=="CNA")
CNA.moth51

# Run adonis
doadonis(physeq = CNA.moth51,category = "Date")


````

### cluster without taxonomy
```{r}
  mothur_slowclust<-mothur_import_and_clean(sharedfile = "mothur_slowclust.shared",taxfile = "mothur_slowclust.taxonomy",mapfile = "map_nomock.txt")

mothur_slowclust
```


The distribution of sample read counts looks the same
```{r,echo=FALSE,warning=FALSE}
# Histogram of sample read counts
ggplot(data.frame(sum=sample_sums(mothur_slowclust)),aes(sum)) + geom_histogram(colour="white",fill="indianred")+ggtitle("Sample read counts (mothur slowclust)")+xlab("total sequences")

```

Now we have even more firmicutes with a few more planctomycetes and verrucos
```{r,fig.height=8,fig.width=12}
make_barplot(physeq = mothur_slowclust,title = "Mothur cluster without taxonomy")
```

```{r}
# Subset to just whole community fraction 
CNA.moth_slowclust <- subset_samples(mothur_slowclust,Fraction=="CNA")
CNA.moth_slowclust

# Run adonis
doadonis(physeq = CNA.moth_slowclust,category = "Date")


````


## UPARSE

Now we will use the UPARSE pipeline to cluster OTUs. One of the big differences between UPARSE and mothur is that UPARSE clusters OTUs before it assigns any taxonomy to the sequences. Taxonomy is assigned after clustering just to the representative sequence from each OTU rather unlike mothur where taxonomy is assigned to each sequence and then the majority classification is given to the OTU. We used the same mothur function classify.seqs() to assign taxonomy to the UPARSE OTUs

### Cutoff 80
First, we will use an assignment cutoff of 80
```{r,echo=FALSE}

uparse80<-uparse_import_and_clean(otufile = "uparse.otutable",taxfile = "uparse80_nr.taxonomy",mapfile = "metadata.txt")

uparse80

```
This histogram looks slightly different than the mothur output but very comparable. We still have many samples with very low read counts due to a high number of chloroplast reads
```{r,echo=FALSE,warning=FALSE}
# Histogram of sample read counts
ggplot(data.frame(sum=sample_sums(uparse80)),aes(sum)) + geom_histogram(colour="white",fill="indianred")+ggtitle("Sample read counts (uparse80)")+xlab("total sequences")

```

This barplot looks VERY different than the mothur output. In the full community and free-living fractions we have far fewer unclassifieds. We have many more verrucos and planctomycetes, which we would expect for this community. 

```{r,fig.height=8,fig.width=12}
make_barplot(uparse80,"Uparse 80")

```

```{r}
# Subset to just whole community fraction 
CNA.uparse80 <- subset_samples(uparse80,Fraction=="CNA")
CNA.uparse80

# Run adonis
doadonis(physeq = CNA.uparse80,category = "Date")


````

### Cutoff 60
Now we will use the same UPARSE otu output but with taxonomy assigned with a 60% cutoff
```{r,echo=FALSE}

uparse60<-uparse_import_and_clean(otufile = "uparse.otutable",taxfile = "Uparse60_noprobs_nr.taxonomy",mapfile = "metadata.txt")

uparse60
```

My Uparse object has 7720 OTUs across 183 samples


```{r,echo=FALSE,warning=FALSE,message=FALSE}
# Histogram of sample read counts
ggplot(data.frame(sum=sample_sums(uparse60)),aes(sum)) + geom_histogram(colour="white",fill="indianred")+ggtitle("Sample read counts (uparse60)")+xlab("total sequences")

```

Using a cutoff of 60, we aren't able to classify many more of our OTUs. This is because we have one OTU -- Otu7 -- that accounts for a large majority of our unclassifieds in the Algal-associated fraction. If you BLAST this OTU in NCBI there really is nothing very closely related to it in the database. 

```{r,echo=FALSE,fig.height=8,fig.width=12}
make_barplot(physeq = uparse60,title = "Uparse 60")
```

```{r}
# Subset to just whole community fraction 
CNA.uparse60 <- subset_samples(uparse60,Fraction=="CNA")
CNA.uparse60

# Run adonis
doadonis(physeq = CNA.uparse60,category = "Date")

````

## Conclusions


1) Running mothur with cluster.split() instead of cluster() produces many more OTUs. For this data it was about 10% more OTUs. These are likely spurious. If you have access to a lot of memory via flux or some other server, I reccomend using cluster()


2) To get accurate taxonomy classifications, use a cutoff of 80
 

3) The adonis results for all methods were almost identical. In general, I think any OTU-based analyses that employ an abundance-weighted beta diversity metric like Bray-curtis will be highly robust to different otu-clustering methods.


4) The uparse and mothur output looked very similar for Marian's inland lakes. It is unclear why there is such a large discrepancy in this Lake Erie dataset


5) For my samples, uparse ran in under 20 minutes on my 2 processor laptop. Mothur took 8 hours on 30 processors on our flux account. 





## Appendix: Functions

### scale_reads
```{r}

## Function to scale read counts by a given library size "n". In all examples I scale by the smallest library size


scale_reads <- function(physeq,n){
  physeq.scale <- transform_sample_counts(physeq, function(x) {(n*x/sum(x))})
  otu_table(physeq.scale) = floor(otu_table(physeq.scale))
  physeq.scale = prune_taxa(taxa_sums(physeq.scale)>0,physeq.scale)
  return(physeq.scale)
}
```
### order_dates
```{r}
# Function to reorder the Date levels because I suck at dates in R
order_dates <- function(df){
  df$Date <- factor(df$Date, levels = c("5/27/14","6/10/14","6/16/14","6/30/14","7/8/14","7/14/14","7/21/14","7/29/14","8/4/14","8/11/14","8/18/14","8/25/14","9/2/14","9/8/14","9/15/14","9/23/14","9/29/14","10/6/14","10/15/14","10/20/14","10/27/14","11/3/14"))
  return(df)
}
```
### mothur_import_and_clean
```{r}
### Data import
## Function to import mothur data into phyloseq and remove non-bacteria taxa as well as samples from blanks, mock communities and the intensive cruises


mothur_import_and_clean<-function(sharedfile,taxfile,mapfile){
  
  mothurdata = import_mothur(mothur_shared_file = sharedfile,
                             mothur_constaxonomy_file = taxfile )

  # We need to change the taxonomy names
  tax_table(mothurdata)<-cbind(tax_table(mothurdata),row.names(tax_table(mothurdata)))
  colnames(tax_table(mothurdata)) <- c("Kingdom","Phylum","Class","Order","Family","Genus","Species")
  
  #import mapfile
  map = import_qiime_sample_data(mapfile)
  colnames(otu_table(mothurdata))<-map$SampleID
  moth_merge = merge_phyloseq(mothurdata,map)

  #Filter out shit 
  moth_good <- subset_samples(moth_merge,Type=="sample")
  moth_good <- subset_samples(moth_good,Intensive=="n")
  moth_good <- subset_taxa(moth_good, Kingdom == "Bacteria")
  moth_good <- subset_taxa(moth_good, Class != "Chloroplast")
  moth_good <- subset_taxa(moth_good, Family!="mitochondria")
  moth_good <- prune_taxa(taxa_sums(moth_good)>0,moth_good)
  return(moth_good)
}
```
### uparse_import_and_clean
```{r}

## Function to import uparse data into phyloseq and remove non-bacteria taxa as well as samples from blanks, mock communities and the intensive cruises

uparse_import_and_clean <- function(otufile,taxfile,mapfile){
  
  # Uparse import
  otu<-read.csv(otufile,sep="\t")
  tax<-read.csv(taxfile,sep="\t")
  rownames(tax)<-rownames(otu)
  map<-read.csv(mapfile,sep="\t")
  rownames(map)<-map$SampleID
  phyOTU<-otu_table(otu,taxa_are_rows=TRUE)
  tax<-as.matrix(tax)
  phyTax<-(tax_table(tax))
  phySam<-sample_data(map)
  uparse.physeq<-phyloseq(phyOTU,phyTax,phySam)
  
  # Filter out bad shit
  uparse.good<-subset_taxa(uparse.physeq,
                           Class!="Chloroplast" & 
                             Kingdom =="Bacteria" & 
                             Family!="mitochondria")
  uparse.good <- subset_samples(uparse.good,Type=="sample" & Intensive=="n")
  uparse.good <- prune_taxa(taxa_sums(uparse.good)>0,uparse.good)
  
  return(uparse.good)
  
}
```
### make_barplot
```{r}
## Function to make a stacked barplot of community composition from a phyloseq object. 
# This function only plots phyla > 1% of the community in each sample

make_barplot<-function(physeq,title){
  # Agglomerate all otu's by phylum level
  physeq_phylum<-tax_glom(physeq,taxrank = "Phylum")

  # Create a new phyloseq object which for each sample only includes phyla with a relative abundance above 1 % 
  physeq_phylum.99 = transform_sample_counts(physeq_phylum, 
                                              function(x){x/sum(x)})
  otu_table(physeq_phylum.99)[otu_table(physeq_phylum.99)<.01] <- 0 
  physeq_phylum.99 = prune_taxa(taxa_sums(physeq_phylum.99)>0,
                                 physeq_phylum.99)

  # Melt into long format and sort by samplenum and phylum
  physeq_phylum.long<-psmelt(physeq_phylum.99)
  physeq_phylum.long<-arrange(physeq_phylum.long,Samplenum,Phylum)

  # Set phylum plotting colors
  phylum.colors <- c(Acidobacteria="darkorchid3",Actinobacteria="#0099ff",Armatimonadetes="orange", Bacteroidetes="#0c2b57",Chlamydiae="#377eb8",Chlorobi="darkgreen", Chloroflexi="magenta",Cyanobacteria="#7dc107","Deinococcus-Thermus"="darkorange",Firmicutes="yellow",Fusobacteria="magenta",Gemmatimonadetes="maroon",Nitrospirae="royalblue",Planctomycetes="cyan",Proteobacteria="#E12C4D", unclassified="grey", Verrucomicrobia="#AD6F3B","white","black","green","blue")

  # Reorder factor levels for dates 
  physeq_phylum.long<-order_dates(physeq_phylum.long)

  # Reorder factor levels for Fraction and Station. Change Fraction factor names
 physeq_phylum.long$Fraction <- factor(physeq_phylum.long$Fraction,
                                       levels=c("CNA","100L","22NA"))
  levels(physeq_phylum.long$Fraction) <- c("Full Community", 
                                           "Particle/Algal", "Free Living")
  physeq_phylum.long$Station <-factor(physeq_phylum.long$Station,
                                      levels=c("WE2","WE12","WE4"))


  # GGPLOT it
  ggplot(physeq_phylum.long,aes(x=Date,y=Abundance,fill=Phylum)) +
        facet_grid(Fraction~Station,scales="free",space="free")+
        geom_bar(stat="identity",position="fill")+ 
   scale_y_continuous(labels = percent_format())+ 
  scale_fill_manual(values = phylum.colors,name="Phylum") + 
    theme(axis.title.x = element_text(size=16,face="bold"),
          axis.text.x = element_text(angle=50, colour = "black", vjust=1, hjust = 1, size=6.5),
          axis.text.y = element_text(colour = "black", size=9),
          axis.title.y = element_text(face="bold", size=16),
          plot.title = element_text(size = 18),
          legend.title = element_text(size=14),
          legend.text = element_text(size = 13),
          legend.position="right")+
    guides(fill = guide_legend(reverse= TRUE,keywidth=1,keyheight=1))+
    xlab("Sampling Date")+
    ylab("Relative Abundance (taxa>.1%)")+
    ggtitle(title)
        
}
```
### doadonis
```{r}
# Function to run adonis test on a distance object and a categorical variable from a metadata data frame
doadonis <- function(physeq, category) {
  physeq.scale <- scale_reads(physeq,min(sample_sums(physeq)))
  bdist<-phyloseq::distance(physeq.scale,"bray")
  col = as(sample_data(physeq),"data.frame")[,category]
  adonis.bdist=adonis(bdist~col)
  print(adonis.bdist) 
  paste("The Homogeneity of dispersion \n")
 

 # Homogeneity of dispersion test
  betatax=betadisper(bdist,col)
  p=permutest(betatax)
  print("The homogeneity of dispersion:") 
  print(p$tab)
}  


```

